### DISCUSSION TITLE
Visual Studio Code Extension for MPRemote

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#11614] [Show and tell]
If you're a fan of VS Code for development work, I have an extension that provides command palette access to a lot of MPRemote's functionality. I've recently made some improvements to the [code](https://github.com/DavesCodeMusings/mpremote-vscode) and [documentation](https://github.com/DavesCodeMusings/mpremote-vscode/wiki).

You may want to check it out if:
* You prefer point and click to typing command-line options
* You like Thonny, but really wish it had git functionality

https://marketplace.visualstudio.com/items?itemName=DavesCodeMusings.mpremote

I also have a VS Code extension for esptool so you can flash devices using point and click dialogs instead of command-line.

https://marketplace.visualstudio.com/items?itemName=DavesCodeMusings.esptool

**Comment by Wind-stormger** [comment]
Hello, can you add graphical buttons and intuitive list of files that can be displayed in the sidebar of VScode?

**Reply by DavesCodeMusings** [reply]
I'll be honest and tell you I have no idea how to do either of those things at the moment. It looks like the buttons might be possible, judging from [this task runner extension](https://marketplace.visualstudio.com/items?itemName=SanaAjani.taskrunnercode) that does something similar. I think the files would be more difficult, as I'm not sure how it could be kept in sync with the flash file system. If you or anyone else has any experience with either of these, I'm open to ideas and/or links to sample code.

So the answer is, maybe, but it's going to take some time.

**Reply by DavesCodeMusings** [reply]
No buttons, but I did manage to add context menus for some of the commands where it makes sense. For example, right-click in the editor window and you get the option to `run` or `cp` (upload) the file. You can also right-click the Explorer pane and have the option to `sync` the local directory tree to the flash filesystem (recursive copy.)

All commands are still available from the command palette (CTRL+SHIFT+P)

Editor window context menu:
![image](https://github.com/micropython/micropython/assets/61114342/480be2b4-898a-4229-881d-a5c7c627f26d)

Explorer pane context menu:
![image](https://github.com/micropython/micropython/assets/61114342/acf57855-8205-4491-8091-25dbbb3ba233)

**Reply by Wind-stormger** [reply]
Much better! 😄 

Don't worry, referring to the source code of other open source extension plug-in projects may help you understand.

For example: https://github.com/huizhougit/githd

Or other open source projects that you think are better~

Let's talk about the content of this update:

This one doesn't seem to work with `MPRemote: Change remote working directory (chdir)`, after I select the port and enter the correct absolute path, Enter doesn't respond.

I would like these two items to automatically close the current REPL first when executing `MPRemote: Run file shown in the active editor (run)` and `MPRemote: Upload file to remote filesystem (cp)`. Manually going to the terminal to close is a bit complicated.

**Reply by DavesCodeMusings** [reply]
Can you check the link? It's giving me a 404 error.

`chdir` is a bit of an oddity. It doesn't actually do anything on the remote system, it simply sets the parent path for anything you upload. So if you `chdir` to _/temp_ and then do an upload of _main.py_ in your editor, the file ends up as _/temp/main.py._ It's admittedly a bit clumsy. I think opening a folder in VS Code and then using `sync` is a much better approach. `sync` does a recursive copy of the files in the (VS Code) open folder to the microcontroller's flash filesystem.

As far as closing the REPL automatically, I'm not exactly sure how to do that. I could send a CTRL-X to the terminal, but if the REPL's not open, you'd get an unwanted character. Not exactly sure how I'd detect if it was open or not.

**Reply by Wind-stormger** [reply]
link update

**Reply by Wind-stormger** [reply]
> Not exactly sure how I'd detect if it was open or not.

The easy way to do that is to ignore it, CTRL-X , newline, and continue.

**Comment by DavesCodeMusings** [comment]
_Deleted outdated screenshots to avoid confusion_

**Comment by DavesCodeMusings** [comment]
Recent improvements include a context menu for access to commands by right-clicking the device's serial port name in the Explorer pane (lower left corner of screenshot.) 

![image](https://github.com/micropython/micropython/assets/61114342/f9623263-fe25-4bdc-9cdf-8659cfcb74bb)

You can find out more in the wiki: https://github.com/DavesCodeMusings/mpremote-vscode/wiki

**Comment by Gogin79** [comment]
Hello, I have some problems with this extension. I had to switched from windows to GNU/Linux from many reasons. I use Debian with LXDE eviroment. I installed vscode via snap store a mpremote via pipx because it is the right and security way for now how mpremote install to Debian. Because -> https://packages.debian.org/unstable/utils/micropython-mpremote . Problem is that mpremote:command in vscode, for example calling command to the REPL -> python3 -m mpremote connect /dev/ttyUSB0 repl and it is right but on my PC it is wrong. In case when I installed mpremote using pipx so the command for using mpremote is just only one word "mpremote". It is posible to fix it please. Ofcourse I will looking for another solution which VScode offers. Excus my English please. I hoop you understood.
![mpremote1](https://github.com/micropython/micropython/assets/15175531/e1236cd2-121d-48ad-a92c-87587f4fbe03)
![mpremote2](https://github.com/micropython/micropython/assets/15175531/3b58b083-f983-4775-b0ee-b8ca86475737)

**Reply by DavesCodeMusings** [reply]
Based on the message: _No module named mpremote_, I'm guessing it's something with the module path.

Maybe try installing from a shell prompt with the command: `python3 -m pip install mpremote`

I do all my MicroPython work on Windows, so I'm not sure if this will solve it or not. And, by the way, your English is fine.

**Reply by Gogin79** [reply]
Thank you for your reply and praise for my english. Problem is that Debian has python intengreted and is approved for use in Debian. And with pip packages it is the same. If it is not approve package to use in the python of Debian distro so one solution is create virtual enviroment and all non-approve package is posible use in this virtual enviroment but it is not function in vscode with mpremote or i dont not know how doit. I have solution now. I installed mpremote with hard method :-) with paremeter --break-system-packages and now all is working. It is Great. Explains everything attached picture. Maybe there will be some consequences in Debian and meybe not.
![mpremote3](https://github.com/micropython/micropython/assets/15175531/a09ebaab-0aea-49ba-b949-17954e77ea0d)

**Reply by DavesCodeMusings** [reply]
I may be able to add a configuration option for situations like this. Something like _mpremote path_ where it defaults to _python3 -m mpremote_, but is user configurable. In your case, I think you'd set it to: _mpremote_ or _/usr/bin/mremote_

It might take a while for me to get to it. Let me know if you think that would help. If yes, I'll put it on the issues list as an enhancement.

**Reply by andrewleech** [reply]
@DavesCodeMusings that sounds like a good option to me, though I'd also suggest you could have it auto-fallback, eg. Try running `python3 -m mpremote` and if that fails just run `mpremote` and let the user ensure it's on the path somewhere.

What @Gogin79 reports has become standard practice on many linux distributions, where they have become quite forceful in trying to prevent users from pip installing packages directly into the system. I personally hate this annoying nag, but I guess they've got their reasons. Regardless, using tools like `pipx` to install python tools like mpremote is definitely gaining popularity!

I'm curious when you say you use Windows primarily, do you run it with `python3 -m mpremote` there too? I guess this works _if_ the user has the microsoft store version of python installed, but would also generally fail if they've installed the official python packages which use `py` instead.

**Reply by DavesCodeMusings** [reply]
On windoze, it's `py.exe -m mpremote`. I download from python.org and install using the defaults, which puts py.exe in the system path.

I have a function in the extension that tries to figure out the executable name based on OS.

```js
export function getPythonExecutableName() {
    let pythonBinary = 'py.exe';  // Assume this is a Windows system for now.
    if (process.platform !== 'win32') {  // win32 is returned for 64-bit OS as well
        pythonBinary = 'python3';
    }
    console.debug('Using Python executable:', pythonBinary);
    return pythonBinary;
}
```

**Reply by andrewleech** [reply]
Ah, so in that case my comment is true but in reverse - it should work with windows when the python.org dist is installed, but fail if the microsoft store version is used (it's all just such a mess). Again, if this falled back on failure to just use `mpremote` that could work on any platform if the path is setup appropriatly.

**Reply by Gogin79** [reply]
If mpremote will install via pipx. So anywhere you write command  "mpremote" to the terminal is does not matter whatever folder you are in will start mpremote. All what is installed via pipx is starting from  virtual enviroment. PATHs are obvious from image.

![mpremote4](https://github.com/micropython/micropython/assets/15175531/7a91459c-1953-429e-b4d7-aa334129f157)

@DavesCodeMusings 
It would be great if you can this condition implented to the VS code. Another users will not have to solve this problem in future.

**Reply by Gogin79** [reply]
Another possibilties is use this https://github.com/dhylands/rshell . But I am very happy with extension MPRemote. Thank you @DavesCodeMusings .

**Comment by DavesCodeMusings** [comment]
_A Windows user, a Mac user, and a Linux user walk into a bar. The bartender says..._

According to the [doc for installing mpremote](https://docs.micropython.org/en/latest/reference/mpremote.html) it looks like Linux users can just type `mpremote`. Installing with pip or pipx doesn't seem to matter. Can I get a Linux user to confirm that?

Also, is there a Mac user who can do the same? Just type `mpremote` at a shell prompt and see if it works.

If this method works for both Mac and Linux, I think it's safe to just use `mpremote` to run commands on those operating systems and skip the whole `python3 -m mpremote` thing.

**Reply by Gogin79** [reply]
@DavesCodeMusings It does not matter how to install mpremote. Pip or pipx. Just type mpremote and all working. I tested it both variant.

**Reply by DavesCodeMusings** [reply]
This is Debian Linux, correct?

**Reply by Gogin79** [reply]
Yesss! :-)

**Reply by DavesCodeMusings** [reply]
Excellent. Now I just need to find a Mac user.

**Reply by ricksorensen** [reply]
On Linux (Ubuntu 22.04) I install  with `pip install mpremote` in a development  python virtual environment.  A `mpremote` binary is installed in to the python environment `.../bin` directory which I can then execute directly.  
For general use I link that binary into my user `bin`   directory which is alway on my path.

**Reply by andrewleech** [reply]
I just tested this on my partners macbook: 
* installed official python 3.11 from python.org
* open terminal
```
mba:~ $ pip3 install mpremote
... normal pip install stuff
mba:~ $ mpremote --version
mpremote 1.21.0
```

**Reply by DavesCodeMusings** [reply]
Sweet! Thanks everyone. I think I have what I need to correctly invoke `mpremote` on Linux, Mac, and Windows. It'll take a while to rework the extension, but at least I know the direction to take.

**Reply by DavesCodeMusings** [reply]
I've just published v1.21.8 of the extension to hopefully make this work better. I use VS Code on Windows, so I can tell you the changes didn't break anything there. As to the effect for Linux and Mac users, I'm hoping I can get some folks with those OS to reply with your results.

For those interested in the details, this is how the extension determines how to call mpremote:

```
export function getMPRemoteName() {
    let mpremote = 'python3 -m mpremote';
    switch (process.platform) {
        case 'win32':  // win32 is returned for 64-bit OS as well
            mpremote = 'py.exe -m mpremote';
            break;
        case 'linux':
        case 'darwin':
            mpremote = 'mpremote';
            break;
    }
    console.debug('Calling mpremote as:', mpremote);
    return mpremote;
```

**Reply by Gogin79** [reply]
Hello @DavesCodeMusings . On Linux (Debian) everything works fine. Good job. Thank you.
![mpremote5](https://github.com/micropython/micropython/assets/15175531/a3603f9f-ead3-4552-91cc-6b29b191d260)

**Reply by DavesCodeMusings** [reply]
Thanks for testing.
